lib/model/model.go

```diff
@@ -784,6 +787,40 @@ func (m *Model) NeedFolderFiles(folder string, page, perpage int) ([]db.FileInfo
 	return progress, queued, rest, total
 }
 
+// RemoteNeedFolderFiles returns paginated list of currently needed files in
+// progress, queued, and to be queued on next puller iteration, as well as the
+// total number of files currently needed.
+func (m *Model) RemoteNeedFolderFiles(device protocol.DeviceID, folder string, page, perpage int) ([]db.FileInfoTruncated, int, error) {
+	m.pmut.RLock()
+	m.fmut.RLock()
+	if err := m.checkDeviceFolderConnectedLocked(device, folder); err != nil {
+		m.fmut.RUnlock()
+		m.pmut.RUnlock()
+		return nil, 0, err
+	}
+	rf := m.folderFiles[folder]
+	m.fmut.RUnlock()
+	m.pmut.RUnlock()
+
+	files := make([]db.FileInfoTruncated, 0, perpage)
+	skip := (page - 1) * perpage
+	get := perpage
+	total := 0
+	rf.WithNeedTruncated(device, func(f db.FileIntf) bool {
+		total++
+		if skip > 0 {
+			skip--
+			return true
+		}
+		if get > 0 {
+			files = append(files, f.(db.FileInfoTruncated))
+		}
+		return true
+	})
+
+	return files, total, nil
```

Comment:
Don't you have the total in the needCount from completion?
In which case you could decrement get every time you append a file, and `return get > 0` to stop iteration when you've got enough.

I guess if this list of out of sync keeps changing, pagination has the nice affect of readjusting the pages count every time you switch the page, yet I suspect this might be sluggish for large amounts of files.